sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala

```diff
@@ -59,6 +61,159 @@ class BucketedReadSuite extends QueryTest with SQLTestUtils with TestHiveSinglet
     }
   }
 
+  // To verify if the bucket pruning works, this function checks two conditions:
+  //   1) Check if the pruned buckets (before filtering) are empty.
+  //   2) Verify the final result is the same as the expected one
+  private def checkPrunedAnswers(
+      bucketSpec: BucketSpec,
+      bucketValues: Seq[Integer],
+      bucketedDataFrame: DataFrame,
+      expectedAnswer: DataFrame): Unit = {
+
+    val BucketSpec(numBuckets, bucketColumnNames, _) = bucketSpec
+    // Limit: bucket pruning only works when the bucket column has one and only one column
+    assert(bucketColumnNames.length == 1)
+    val bucketColumnIndex = bucketedDataFrame.schema.fieldIndex(bucketColumnNames.head)
+    val bucketColumn = bucketedDataFrame.schema.toAttributes.head
+    val matchedBuckets = new BitSet(numBuckets)
+    bucketValues.foreach { value =>
+      matchedBuckets.set(DataSourceStrategy.getBucketId(bucketColumn, numBuckets, value))
+    }
+
+    // Filter could hide the bug in bucket pruning. Thus, skipping all the filters
+    val rdd = bucketedDataFrame.queryExecution.executedPlan.find(_.isInstanceOf[PhysicalRDD])
+    assert(rdd.isDefined)
+
+    val checkBucketId = rdd.get.execute().map(_.copy()).mapPartitions(iter => {
+      iter.map(row =>
+        DataSourceStrategy.getBucketId(
+          bucketColumn, numBuckets, row.get(bucketColumnIndex, bucketColumn.dataType)))})
+    // Check if all the returned rows are from the non-pruned buckets
+    assert(checkBucketId.collect().forall(matchedBuckets.get))
+
+    checkAnswer(
+      expectedAnswer
+        .orderBy(expectedAnswer.logicalPlan.output.map(attr => Column(attr)) : _*),
```

Comment:
There is a test case which is using a different Dataframe `nullStrings`. Thus, the schema is different. If you want, I can create a dataFrame with the same schema for this test case. Let me know if you want me to do the change. Thanks!